server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java

```diff
@@ -207,7 +207,11 @@ void checkLimits(IndexSettings settings) {
     }
 
     private void checkFieldLimit(long limit) {
-        if (fieldMappers.size() + objectMappers.size() - mapping.getSortedMetadataMappers().length > limit) {
+        checkFieldLimit(limit, 0);
+    }
+
+    public void checkFieldLimit(long limit, int additionalFieldsToAdd) {
+        if (fieldMappers.size() + objectMappers.size() + additionalFieldsToAdd - mapping.getSortedMetadataMappers().length > limit) {
             throw new IllegalArgumentException("Limit of total fields [" + limit + "] has been exceeded");
```

Comment:
I wonder  if we can be more helpful with our messages here. 
There's 2 scenarios:
1) "The straw that broke the camel's back" - i.e.  there was steady growth towards the field limit over many docs.
2) A rogue document.

For 1) maybe we  could log warnings as we get close to hitting the limit - kind of like disk space threshold warnings. This would allow users the ability to react/redesign prior  to disaster.
For 2) maybe we should quantify how many _new_ fields a single document was attempting to add.

Reply:
I've expanded the message to cover scenario 2 in 8bda5d8.
I'm not sure about how common scenario 1 is. I haven't seen cases where this limit was slowly reached. I'm doubtful if logging would help, perhaps some kind of monitoring with Kibana UI integration would help. Or part of a more general "cluster health" API. Not something I would like to address as part of this PR.